internal/mesh/internal/controllers/routes/generate.go

```diff
@@ -0,0 +1,631 @@
+			if backendRef.BackendRef.Port == "" {
+				backendRef.BackendRef.Port = port
+			}
+
+			var (
+				backendTarget string
+				backendSvc    = serviceGetter.GetService(backendRef.BackendRef.Ref)
+			)
+			if shouldRouteTrafficToBackend(backendSvc, backendRef.BackendRef) {
+				details := &pbmesh.BackendTargetDetails{
+					BackendRef: backendRef.BackendRef,
+				}
+				backendTarget = node.AddTarget(backendRef.BackendRef, details)
+			} else {
+				backendTarget = types.NullRouteBackend
+			}
+
+			ibr := &pbmesh.ComputedGRPCBackendRef{
+				BackendTarget: backendTarget,
+				Weight:        backendRef.Weight,
+				Filters:       backendRef.Filters,
+			}
+			irule.BackendRefs = append(irule.BackendRefs, ibr)
+		}
+
+		node.GRPCRules = append(node.GRPCRules, irule)
+	}
+
+	return node
+}
+
+func compileTCPRouteNode(
+	port string,
+	res *pbresource.Resource,
+	route *pbmesh.TCPRoute,
+	serviceGetter serviceGetter,
+) *inputRouteNode {
+	route = protoClone(route)
+
+	node := newInputRouteNode(port)
+	dec := &types.DecodedTCPRoute{
+		Resource: res,
+		Data:     route,
+	}
+
+	node.RouteType = types.TCPRouteType
+	node.TCP = dec
+	node.TCPRules = make([]*pbmesh.ComputedTCPRouteRule, 0, len(route.Rules))
+	for _, rule := range route.Rules {
+		irule := &pbmesh.ComputedTCPRouteRule{
+			BackendRefs: make([]*pbmesh.ComputedTCPBackendRef, 0, len(rule.BackendRefs)),
+		}
+
+		// https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1alpha2.TCPRoute
+
+		for _, backendRef := range rule.BackendRefs {
+			// Infer port name from parent ref.
+			if backendRef.BackendRef.Port == "" {
+				backendRef.BackendRef.Port = port
+			}
+
+			var (
+				backendTarget string
+				backendSvc    = serviceGetter.GetService(backendRef.BackendRef.Ref)
+			)
+			if shouldRouteTrafficToBackend(backendSvc, backendRef.BackendRef) {
+				details := &pbmesh.BackendTargetDetails{
+					BackendRef: backendRef.BackendRef,
+				}
+				backendTarget = node.AddTarget(backendRef.BackendRef, details)
+			} else {
+				backendTarget = types.NullRouteBackend
+			}
+
+			ibr := &pbmesh.ComputedTCPBackendRef{
+				BackendTarget: backendTarget,
+				Weight:        backendRef.Weight,
+			}
+			irule.BackendRefs = append(irule.BackendRefs, ibr)
+		}
+
+		node.TCPRules = append(node.TCPRules, irule)
+	}
+
+	return node
+}
+
+func shouldRouteTrafficToBackend(backendSvc *types.DecodedService, backendRef *pbmesh.BackendReference) bool {
+	if backendSvc == nil {
+		return false
+	}
+
+	var (
+		found  = false
+		inMesh = false
+	)
+	for _, port := range backendSvc.Data.Ports {
+		if port.Protocol == pbcatalog.Protocol_PROTOCOL_MESH {
+			inMesh = true
+			continue
+		}
+		if port.TargetPort == backendRef.Port {
+			found = true
+		}
+	}
+
+	return inMesh && found
+}
+
+func createDefaultRouteNode(
+	parentServiceRef *pbresource.Reference,
+	port string,
+	typ *pbresource.Type,
+) *inputRouteNode {
+	// always add the parent as a possible target due to catch-all
```

Comment:
This comment still reflects _part_ of the code, but generally speaking if you have a route and you don't match some of the traffic then the remainder of the traffic is null routed, instead of being directed to a catch all.

I updated the surrounding code to add in the null route, so at worst this code still adds the extra (extraneous) self target, but does nothing with it.